precompiles/referenda/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+	BalanceOf<Runtime>,
+	<Runtime as frame_system::Config>::BlockNumber,
+>>::Id;
+type BoundedCallOf<Runtime> = Bounded<<Runtime as pallet_referenda::Config>::RuntimeCall>;
+
+type OriginOf<Runtime> =
+	<<Runtime as frame_system::Config>::RuntimeOrigin as OriginTrait>::PalletsOrigin;
+
+/// A precompile to wrap the functionality from pallet-referenda.
+pub struct ReferendaPrecompile<Runtime, GovOrigin: TryFrom<u16>>(PhantomData<(Runtime, GovOrigin)>);
+
+#[precompile_utils::precompile]
+impl<Runtime, GovOrigin> ReferendaPrecompile<Runtime, GovOrigin>
+where
+	Runtime: pallet_referenda::Config + pallet_evm::Config + frame_system::Config,
+	OriginOf<Runtime>: From<GovOrigin>,
+	<Runtime as frame_system::Config>::RuntimeCall:
+		Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<<Runtime as frame_system::Config>::RuntimeCall as Dispatchable>::RuntimeOrigin:
+		From<Option<Runtime::AccountId>>,
+	<Runtime as frame_system::Config>::RuntimeCall: From<ReferendaCall<Runtime>>,
+	Runtime::BlockNumber: Into<U256>,
+	TrackIdOf<Runtime>: TryFrom<u16> + TryInto<u16>,
+	BalanceOf<Runtime>: Into<U256>,
+	GovOrigin: TryFrom<u16>,
+{
+	// The accessors are first. They directly return their result.
+	#[precompile::public("referendumCount()")]
+	#[precompile::view]
+	fn referendum_count(handle: &mut impl PrecompileHandle) -> EvmResult<U256> {
+		// Fetch data from pallet
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let ref_count = ReferendumCount::<Runtime>::get();
+		log::trace!(target: "referendum-precompile", "Referendum count is {:?}", ref_count);
+
+		Ok(ref_count.into())
+	}
+
+	#[precompile::public("submissionDeposit()")]
+	#[precompile::view]
+	fn submission_deposit(handle: &mut impl PrecompileHandle) -> EvmResult<U256> {
+		// Fetch data from pallet
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let submission_deposit = Runtime::SubmissionDeposit::get();
+		log::trace!(target: "referendum-precompile", "Submission deposit is {:?}", submission_deposit);
+
+		Ok(submission_deposit.into())
+	}
+
+	#[precompile::public("decidingCount(uint16)")]
+	#[precompile::view]
+	fn deciding_count(handle: &mut impl PrecompileHandle, track_id: u16) -> EvmResult<U256> {
+		// Fetch data from pallet
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let track_id: TrackIdOf<Runtime> = track_id
+			.try_into()
+			.map_err(|_| RevertReason::value_is_too_large("Track id type").into())
+			.in_field("trackId")?;
+		let deciding_count = DecidingCount::<Runtime>::get(track_id);
+		log::trace!(
+			target: "referendum-precompile", "Track {:?} deciding count is {:?}",
+			track_id,
+			deciding_count
+		);
+
+		Ok(deciding_count.into())
+	}
+
+	#[precompile::public("trackIds()")]
+	#[precompile::view]
+	fn track_ids(handle: &mut impl PrecompileHandle) -> EvmResult<Vec<u16>> {
+		// Fetch data from runtime
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
```

Comment:
I don't think this comes from storage. Also, it would not account for the scaling per number of tracks.

But maybe those two things cancel each other out, at least for now :)

Reply:
Indeed, tracks info are static hardcoded data, I think we should not account any gas cost here

Reply:
We need to account for some, otherwise you could have a malicious contract do a loop over `track_ids()` to create spam. But it probably doesn't need to be much

Reply:
> Also, it would not account for the scaling per number of tracks. But maybe those two things cancel each other out, at least for now :)

Yes iterating over tracks has some cost. I'll make a follow up ticket. The accurate approach is to benchmark the iteration and charge that weight => gas instead of charging 1 db read.